jwql/instrument_monitors/pipeline_tools.py

```diff
@@ -0,0 +1,249 @@
+                         'saturation': SaturationStep, 'superbias': SuperBiasStep}
+
+# Readout patterns that have nframes != a power of 2. These readout patterns
+# require the group_scale pipeline step to be run.
+GROUPSCALE_READOUT_PATTERNS = ['NRSIRS2']
+
+
+def completed_pipeline_steps(filename):
+    """
+    Return a list of the completed pipeline steps for a given file.
+
+    Parameters
+    ----------
+    filename : str
+        File to examine
+
+    Returns
+    -------
+    completed : collections.OrderedDict
+        Dictionary with boolean entry for each pipeline step,
+        indicating which pipeline steps have been run on filename
+    """
+    # Initialize using PIPE_KEYWORDS so that entries are guaranteed to
+    # be in the correct order
+    completed = OrderedDict({})
+    for key in PIPE_KEYWORDS.values():
+        completed[key] = False
+
+    header = fits.getheader(filename)
+    for key in PIPE_KEYWORDS.keys():
+        value = header.get(key)
+        if value == 'COMPLETE':
+            completed[PIPE_KEYWORDS[key]] = True
+    return completed
+
+
+def get_pipeline_steps(instrument):
+    """Get the names and order of the calwebb_detector1
+    pipeline steps for a given instrument. Use values that match up with the values in the
+    PIPE_STEP defintion in definitions.py
+
+    Parameters
+    ----------
+    instrument : str
+        Name of JWST instrument
+
+    Returns
+    -------
+    steps : collections.OrderedDict
+        Dictionary of step names (and modules? do we care?)
+    """
+    instrument = instrument.upper()
+    if instrument not in JWST_INSTRUMENT_NAMES_UPPERCASE.values():
+        raise ValueError("WARNING: {} is not a valid instrument name.".format(instrument))
+    # all_steps = Detector1Pipeline.step_defs
+
+    # Order is important in 'steps' lists below!!
+    if instrument == 'MIRI':
+        steps = ['group_scale', 'dq_init', 'saturation', 'ipc', 'firstframe', 'lastframe',
+                 'linearity', 'rscd', 'dark_current', 'refpix', 'persistence', 'jump', 'rate']
+        # No persistence correction for MIRI
+        steps.remove('persistence')
+        # MIRI is limited to one frame per group
+        steps.remove('group_scale')
+    else:
+        steps = ['group_scale', 'dq_init', 'saturation', 'ipc', 'superbias', 'refpix', 'linearity',
+                 'persistence', 'dark_current', 'jump', 'rate']
+
+        # No persistence correction for NIRSpec
+        if instrument == 'NIRSPEC':
+            steps.remove('persistence')
+        else:
+            # NIRCam, NISISS, FGS all do not need group scale as nframes is
+            # always a multiple of 2
+            steps.remove('group_scale')
+
+    # IPC correction currently not done for any instrument
+    steps.remove('ipc')
+
+    # Initialize using PIPE_KEYWORDS so the steps will be in the right
+    # order
+    req = OrderedDict({})
```

Comment:
@bhilbert4 I think you mean "required" by `req` here.  It might be more readable to use `required_steps` or something like that as a variable name instead of `req`, otherwise it isn't entirely clear what `req` means.

Reply:
Updated to `required_steps`